src/compiler/checker.ts

```diff
@@ -15666,11 +15674,22 @@ namespace ts {
             return !!(getGenericObjectFlags(type) & ObjectFlags.IsGenericIndexType);
         }
 
+        function isUnionWithGenericDiscriminantProperty(type: UnionType) {
+            getPropertiesOfType(type); // initialize property cache (includes partial props)
+            const props = arrayFrom((type.propertyCache || emptySymbols).values());

```

Comment:
what does this give you that getPropertiesOfType doesn't? Partial props? Why is that important?

Reply:
Yeah, partial props. Because partial props in unions with fresh object literals may still discriminate.